src/mesh.cpp

```diff
@@ -1282,6 +1323,77 @@ openmc_mesh_set_params(int32_t index, int n, const double* ll, const double* ur,
   return 0;
 }
 
+//! Get the rectilinear mesh grid
+extern "C" int
+openmc_rectilinear_mesh_get_grid(int32_t index, double** grid_x, int* nx,
+                       double** grid_y, int * ny, double** grid_z, int* nz)
+{
+  if (int err = check_mesh_type(index, "rectilinear")) return err;
+  RectilinearMesh* m = dynamic_cast<RectilinearMesh*>(model::meshes[index].get());
+
+  if (m->lower_left_.dimension() == 0) {
+    set_errmsg("Mesh parameters have not been set.");
+    return OPENMC_E_ALLOCATE;
+  }
+
+  *grid_x = m->grid_[0].data();
+  *nx = m->grid_[0].size();
+  *grid_y = m->grid_[1].data();
+  *ny = m->grid_[1].size();
+  *grid_z = m->grid_[2].data();
+  *nz = m->grid_[2].size();
+
+  return 0;
+}
+
+//! Set the regular mesh parameters
+extern "C" int
+openmc_rectilinear_mesh_set_grid(int32_t index, const double* grid_x,
+                       const int nx, const double* grid_y, const int ny,
+                       const double* grid_z, const int nz)
+{
+  if (int err = check_mesh_type(index, "rectilinear")) return err;
+  RectilinearMesh* m = dynamic_cast<RectilinearMesh*>(model::meshes[index].get());
+
+  m->n_dimension_ = 3;
+  m->grid_.resize(m->n_dimension_);
+
+  for (int i = 0; i < nx; i++) {
+    m->grid_[0].push_back(grid_x[i]);
+  }
+  for (int i = 0; i < ny; i++) {
+    m->grid_[1].push_back(grid_y[i]);
+  }
+  for (int i = 0; i < nz; i++) {
+    m->grid_[2].push_back(grid_z[i]);
+  }
+
+  m->shape_ = {static_cast<int>(m->grid_[0].size()) - 1,
+               static_cast<int>(m->grid_[1].size()) - 1,
+               static_cast<int>(m->grid_[2].size()) - 1};
+
+  for (const auto& g : m->grid_) {
+    if (g.size() < 2) {
+      set_errmsg("x-, y-, and z- grids for rectilinear meshes "
+        "must each have at least 2 points");
+      return OPENMC_E_INVALID_ARGUMENT;
+    }
+    for (int i = 1; i < g.size(); ++i) {
+      if (g[i] <= g[i-1]) {
+        std::cout << g[i] << " " << g[i-1] << "\n";
+        set_errmsg("Values in for x-, y-, and z- grids for "
+          "rectilinear meshes must be sorted and unique.");
+        return OPENMC_E_INVALID_ARGUMENT;
+      }
+    }
+  }
```

Comment:
Maybe we should put this error checking into something like a `RectilinearMesh::set_grid` function. That way, we can use the same error checking for both the C-API and the normal XML input sequence.